### DISCUSSION TITLE
2019 Release

### DISCUSSION THREAD

**Discussion by oschuett** [cp2k/cp2k#2593] [Team Posts]
Traditionally, we made a release at least once a year.
Since the end of 2019 is approaching quickly, I'd like to kick off the discussion.

**Comment by hfp** [comment]
It makes sense to release a new version with a time horizon of ~4 months (perhaps it can slip into 2020). I already got an inquiry from a customer to run CP2K 7.0-dev i.e., with a supposedly good SHA-revision...

Also, CP2K currently keeps accumulating more changes i.e., different behaviour than the last release (even if it was a correction). Though, 7.1 will be already a big departure from v6.1 given an externalized DBCSR, LIBINT 2.x and many other changes.

**Comment by oschuett** [comment]
Do you expect any particular feature to arrive within those 4 months?
I'd prefer to make a release this year and plan the next one for early 2020.

**Comment by hfp** [comment]
No, it was just to avoid saying "asap" and to give time for open issues that are expected to be fixed in v7.1.

**Comment by oschuett** [comment]
I looked through our commits from the last two months. We currently have the following _construction sites_:
  - MP2/RPA/GW (@fstein93)
  - XAS_TDP (@abussy)
  - PINT / Helium (@cschran)
  - SIRIUS (@mtaillefumier)

Maybe you folks could comment on when you'd be comfortable with a release.

Is there anything else we should consider?

**Comment by dev-zero** [comment]
maybe take a look at the list of issues?

**Comment by oschuett** [comment]
Oh, I see you've started to assign [milestone 7.1](https://github.com/cp2k/cp2k/milestone/3) to some issues.

For past releases we have not paid much attention to the bug tracker. Many of our issues are longstanding (e.g. [#578](https://github.com/cp2k/cp2k/issues/578)) and might never get resolved. I'd propose we only wait for issues that affect many users and on which somebody is actively working.

We've already accumulated a great deal of improvements. We shouldn't hold those back from our users unnecessarily.

**Comment by mkrack** [comment]
This will be the first CP2K with github, since the former release was still with svn using sourceforge. There was a kind of release process (or rather a [check list](https://www.cp2k.org/dev:release_checklistl)) which is now partially obsolete and should be updated. Clean regression tests and a very simple scalability test (water benchmark) were the checks required to pass as preliminaries for a release. That is done now more or less continuously. I think a first step would be to set a deadline in the near future for commits entering the new release, e.g. Fri Nov 29.

**Comment by cschran** [comment]
Indeed, we introduced some updates to the PINT and Helium code. 

1. We improved and corrected the implementation of the worm algorithm for bosonic path integral Monte Carlo sampling (based on this [paper](https://doi.org/10.1103/PhysRevLett.96.070601)).
2. We implemented path integral centroid constraints / restraints for thermodynamic integration or umbrella sampling including the quantum nature of the nuclei.
3. Finally, a new path integral coupling scheme for different number of path integral replicas has been introduced for the PIMC and PIMD subsystem in the implemented hybrid MC/MD technique.

The changes are almost final, but we do expect a small update to the worm implementation (1.) for better performance that should be added in the next days.

**Comment by fstein93** [comment]
In the MP2/ RPA/ GW code, there are the following news: (EDITED to keep everything together)
1) Cubic-scaling Laplace-SOS-MP2
2) k-points for low-scaling GW (by @JWilhelm, preliminary version)
3) Improved performance of cubic-scaling RPA and SOS-MP2 calculations (by @pseewald)
4) Singles corrections for RPA (by @rybkinjr)
5) Long-range potential for MP2 and RPA
6) Refactoring of the code, changes in the input (see PR #272, #412)

There is still some ongoing work, but that can be dealt with in a later release!

**Comment by rybkinjr** [comment]
There are also singles corrections. A small but likely useful thing.

**Comment by fstein93** [comment]
Sry Vladimir.
@JWilhelm told me, we should mark that k-point code preliminary.

**Comment by abussy** [comment]
As for the XAS_TDP code, I believe that after PR https://github.com/cp2k/cp2k/pull/606 is merged, it will be ready for the release.

**Comment by pseewald** [comment]
There has been some work from my side (to be finished in the next days) to completely replace the old implementation of cubic scaling RPA/SOS-MP2/GW methods with an easier to use and faster implementation based on DBCSR tensors. The new implementation of cubic scaling RPA can then be declared as stable and officially be announced as part of the release.

It was careless of me to merge in a not yet released version of DBCSR (in cp2k/cp2k#651) so shortly before the release. If we want a CP2K release to be based on a stable version of DBCSR, I suggest the following options:
- wait for DBCSR v2.1.0 which should come beginning of next year
- partially revert cp2k/cp2k#651 in order to base CP2K on DBCSR 2.0.1 (only a few lines need to be reverted)

The RI-implementation of Hartree-Fock exchange (cp2k/cp2k#581) will not be merged before the release since this is still work in progress.

**Comment by oschuett** [comment]
@pseewald, Great work! It's amazing how quickly you managed to migrate RPA+Co to tensors.

If it's not too much of a hassle, I'd prefer to rollback to DBCSR 2.0.1 and have the release still in 2019.

**Comment by dev-zero** [comment]
I don't mind waiting for 2020 to make the next release, especially since lately there have also been more changes which may benefit from more testing.

**Comment by pseewald** [comment]
I opened a pull request cp2k/cp2k#677 to go back to DBCSR 2.0.1 as a preparation for a potential CP2K release this year. We should leave this open til we are ready for the release.

**Comment by oschuett** [comment]
@dev-zero, which concrete changes are you worried about and what additional tests should we, in you opinion, perform?

From what I can tell, everybody has polished their contributions and given their Ok. All dashboard tests are passing.  Hence, I'd say we're as ready as we'll ever be and should therefore release today's https://github.com/cp2k/cp2k/commit/d15fe601a7ac210c742e5c756bfe2ca1dee4d8a0 as `v7.1`.

**Comment by cschran** [comment]
If it's possible, could PR [#680](https://github.com/cp2k/cp2k/pull/680) still be merged before the release (our core-developer is not available at the moment)? It's minimal changes to the code, but would allow correct output of force_eval properties in PINT runs. Thanks a lot, Christoph

**Comment by oschuett** [comment]
I merged [#680](https://github.com/cp2k/cp2k/pull/680) and [#644](https://github.com/cp2k/cp2k/pull/644). The [#678](https://github.com/cp2k/cp2k/pull/678) looks like a larger change and all other PRs are not ready.

The dashboard is still green. So, let's aim for https://github.com/cp2k/cp2k/commit/8bc402c2628c1c56e286dc7e72b2dc361f6fcb16.

**Comment by fstein93** [comment]
If you think so. To me, it was an addition to #563.

**Comment by oschuett** [comment]
If you deem [#678](https://github.com/cp2k/cp2k/pull/678) ready for release then I'm happy to merge it.

**Comment by dev-zero** [comment]
> @dev-zero, which concrete changes are you worried about and what additional tests should we, in you opinion, perform?

* Multi-GPU systems ... otherwise this has to make it into the release notes' "known issues" section
* `mpirun -np 8 exe/local/cp2k.popt tests/xTB/regtest-3/h2o_rtp.inp`
* checking consistency of our version checks for libraries and what we ship in the toolchain (cp2k/cp2k#562)
* check our documentation wrt recommended/supported versions of libraries and compilers

> From what I can tell, everybody has polished their contributions and given their Ok. All dashboard tests are passing. Hence, I'd say we're as ready as we'll ever be and should therefore release today's cp2k/cp2k@d15fe60 as v7.1.

Can you (or somebody else) please start writing up the release notes to go with the release then?
Updating https://www.cp2k.org/version_history might be a good start, independently of when the release is going to happen exactly.

I've updated the [release checklist](https://www.cp2k.org/dev:release_checklist?s[]=release) to Git, although it is untested until know, except for the automated source tarball generation.

**Comment by oschuett** [comment]
I like the idea of adding a "known issues" section to the release notes. I think, this is the right way to address the fact that there will always be known bugs and not-yet-implemented features.

> Updating https://www.cp2k.org/version_history might be a good start

I went over past commits and drafted a list of keywords. I'd invite everybody to fill in a description of their contributions. 

@dev-zero, thanks for updating the release checklist.

**Comment by oschuett** [comment]
The release is out: https://github.com/cp2k/cp2k/releases/tag/v7.1.0 .
Thanks everybody for contributing and Merry Christmas!

**Comment by hfp** [comment]
I just realized that downloading [`v7.1.0.tar.gz`](https://github.com/cp2k/cp2k/archive/v7.1.0.tar.gz) from [https://github.com/cp2k/cp2k/releases/tag/v7.1.0](https://github.com/cp2k/cp2k/releases/tag/v7.1.0) misses all submodules (exts) namely DBCSR from the source package. The manually attached artifacts such as [`cp2k-7.1.tar.bz2`](https://github.com/cp2k/cp2k/releases/download/v7.1.0/cp2k-7.1.tar.bz2) however contain all necessary bits ("self-contained"). Perhaps it is possible to configure what is to be included in automatically generated archives (`.gitattributes`?).

**Comment by dev-zero** [comment]
This is a known issue and the reason for providing the complete tarball, automatically generated by using GitHub Actions.

**Comment by hfp** [comment]
@dev-zero Thank you, Tiziano!
